# Broadband Access

**Notebook: 1-web-data**

# Abstract

**Purpose:** 
This notebook (intended for a technical audience) pursues a preliminary understanding of data sources and variable labels relevant for exploring disparities in broadband Internet access across lower-level geographies and demographics.

**Acknowledgments:** 
Data source: U.S. Census Bureau, 2019 American Community Survey 1-Year Estimates (https://data.census.gov/cedsci/table?q=broadband&g=0400000US51.050000&y=2019&d=ACS%201-Year%20Estimates%20Data%20Profiles&tid=ACSDP1Y2019.DP02&hidePreview=false)


# Scope (notebook)

 - import libraries
 - read American Community Survey (ACS) CSV data into Pandas
 - inspect data for task-relevant variables
 - subset data
 - serialize objects for further study

# Setup

In [1]:
# importing packages
import os
import pandas as pd

# Obtain

In [2]:
# instantiating the data file as a dataframe
data = pd.read_csv('../data/ACSDP1Y2019.DP02_data_with_overlays_2020-11-07T164043.csv')

In [3]:
# instantiating the metadata file as a dataframe
metadata = pd.read_csv('../data/ACSDP1Y2019.DP02_metadata_2020-11-07T164043.csv')

In [4]:
# viewing dataset info
data.info(), data.shape

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31 entries, 0 to 30
Columns: 614 entries, GEO_ID to DP02_0153PM
dtypes: object(614)
memory usage: 148.8+ KB


(None, (31, 614))

There are 31 rows and 614 columns in this dataset.

In [5]:
# viewing first rows
data.head()

GEO_ID                        NAME  \
0              id        Geographic Area Name   
1  0500000US51003  Albemarle County, Virginia   
2  0500000US51013  Arlington County, Virginia   
3  0500000US51015    Augusta County, Virginia   
4  0500000US51019    Bedford County, Virginia   

                                       DP02_0001E  \
0  Estimate!!HOUSEHOLDS BY TYPE!!Total households   
1                                           43754   
2                                          111887   
3                                           29951   
4                                           31390   

                                          DP02_0001M  \
0  Margin of Error!!HOUSEHOLDS BY TYPE!!Total hou...   
1                                               1134   
2                                               1847   
3                                               1333   
4                                               1264   

                                     DP02_0001PE  \
0  Percent!!HOUSEHOLDS BY TYPE!!Total households   
1                                          43754   
2                                         111887   
3                                          29951   
4                                          31390   

                                         DP02_0001PM  \
0  Percent Margin of Error!!HOUSEHOLDS BY TYPE!!T...   
1                                                (X)   
2                                                (X)   
3                                                (X)   
4                                                (X)   

                                          DP02_0002E  \
0  Estimate!!HOUSEHOLDS BY TYPE!!Total households...   
1                                              22481   
2                                              41865   
3                                              17168   
4                                              18059   

                                          DP02_0002M  \
0  Margin of Error!!HOUSEHOLDS BY TYPE!!Total hou...   
1                                               1665   
2                                               2320   
3                                               1482   
4                                               1613   

                                         DP02_0002PE  \
0  Percent!!HOUSEHOLDS BY TYPE!!Total households!...   
1                                               51.4   
2                                               37.4   
3                                               57.3   
4                                               57.5   

                                         DP02_0002PM  ...  \
0  Percent Margin of Error!!HOUSEHOLDS BY TYPE!!T...  ...   
1                                                3.5  ...   
2                                                2.0  ...   
3                                                4.4  ...   
4                                                4.7  ...   

                                         DP02_0151PE  \
0  Percent!!COMPUTERS AND INTERNET USE!!Total hou...   
1                                              43754   
2                                             111887   
3                                              29951   
4                                              31390   

                                         DP02_0151PM  \
0  Percent Margin of Error!!COMPUTERS AND INTERNE...   
1                                                (X)   
2                                                (X)   
3                                                (X)   
4                                                (X)   

                                          DP02_0152E  \
0  Estimate!!COMPUTERS AND INTERNET USE!!Total ho...   
1                                              40025   
2                                             107630   
3                                              27446   
4                                              27730   

                     

In [6]:
# viewing last rows
data.tail()

GEO_ID                           NAME DP02_0001E DP02_0001M  \
26  0500000US51740      Portsmouth city, Virginia      36089       1133   
27  0500000US51760        Richmond city, Virginia      89878       2207   
28  0500000US51770         Roanoke city, Virginia      40810       1405   
29  0500000US51800         Suffolk city, Virginia      34597       1038   
30  0500000US51810  Virginia Beach city, Virginia     175029       2072   

   DP02_0001PE DP02_0001PM DP02_0002E DP02_0002M DP02_0002PE DP02_0002PM  ...  \
26       36089         (X)      11652       1672        32.3         4.4  ...   
27       89878         (X)      24807       1897        27.6         2.2  ...   
28       40810         (X)      12425       1674        30.4         3.9  ...   
29       34597         (X)      17742       1329        51.3         4.1  ...   
30      175029         (X)      89905       3624        51.4         1.9  ...   

   DP02_0151PE DP02_0151PM DP02_0152E DP02_0152M DP02_0152PE DP02_0152PM  \
26       36089         (X)      33571       1227        93.0         1.7   
27       89878         (X)      81418       2656        90.6         2.0   
28       40810         (X)      35580       1779        87.2         3.2   
29       34597         (X)      32162       1204        93.0         2.3   
30      175029         (X)     169091       2530        96.6         0.9   

   DP02_0153E DP02_0153M DP02_0153PE DP02_0153PM  
26      31644       1221        87.7         2.3  
27      73754       2573        82.1         2.4  
28      32035       1907        78.5         4.3  
29      29058       1309        84.0         3.3  
30     162381       2694        92.8         1.1  

[5 rows x 614 columns]

Each row appears to represent a county or county-equivalent. There are 133 such geographies in Virginia, so only a portion of those are included in the 1-year estimates.

We see that most of the column lables are coded.

The first row appears to contain descriptive metadata. For example, in the column labeled `DP02_0001E`,  "Estimate!!HOUSEHOLDS BY TYPE!!Total households" is the first entry. Let's verify that this matches our `metadata` file.

In [7]:
# viewing first rows from metadata
metadata.head()

GEO_ID                                                 id
0         NAME                               Geographic Area Name
1   DP02_0001E     Estimate!!HOUSEHOLDS BY TYPE!!Total households
2   DP02_0001M  Margin of Error!!HOUSEHOLDS BY TYPE!!Total hou...
3  DP02_0001PE      Percent!!HOUSEHOLDS BY TYPE!!Total households
4  DP02_0001PM  Percent Margin of Error!!HOUSEHOLDS BY TYPE!!T...

The entries match. Let's remove the 0th row from `data` and find `metadata` columns containing the string "broadband".

In [8]:
# deleting the 0th row
update_data = data.drop([data.index[0]]) 

# resetting the index to increment from 0
update_data.reset_index(drop=True, inplace=True)

update_data

GEO_ID                             NAME DP02_0001E DP02_0001M  \
0   0500000US51003       Albemarle County, Virginia      43754       1134   
1   0500000US51013       Arlington County, Virginia     111887       1847   
2   0500000US51015         Augusta County, Virginia      29951       1333   
3   0500000US51019         Bedford County, Virginia      31390       1264   
4   0500000US51041    Chesterfield County, Virginia     128920       1764   
5   0500000US51059         Fairfax County, Virginia     400186       2978   
6   0500000US51061        Fauquier County, Virginia      24270       1052   
7   0500000US51069       Frederick County, Virginia      32010       1398   
8   0500000US51085         Hanover County, Virginia      39521       1078   
9   0500000US51087         Henrico County, Virginia     130591       1988   
10  0500000US51095      James City County, Virginia      30262       1513   
11  0500000US51107         Loudoun County, Virginia     133750       1935   
12  0500000US51121      Montgomery County, Virginia      36244       1423   
13  0500000US51153  Prince William County, Virginia     143508       1907   
14  0500000US51161         Roanoke County, Virginia      37506       1472   
15  0500000US51165      Rockingham County, Virginia      30462       1157   
16  0500000US51177    Spotsylvania County, Virginia      46427       1134   
17  0500000US51179        Stafford County, Virginia      48092       1304   
18  0500000US51199            York County, Virginia      26008        921   
19  0500000US51510        Alexandria city, Virginia      71005       1523   
20  0500000US51550        Chesapeake city, Virginia      86878       1764   
21  0500000US51650           Hampton city, Virginia      55633       1442   
22  0500000US51680         Lynchburg city, Virginia      28950       1238   
23  0500000US51700      Newport News city, Virginia      70918       2279   
24  0500000US51710           Norfolk city, Virginia      88387       2248   
25  0500000US51740        Portsmouth city, Virginia      36089       1133   
26  0500000US51760          Richmond city, Virginia      89878       2207   
27  0500000US51770           Roanoke city, Virginia      40810       1405   
28  0500000US51800           Suffolk city, Virginia      34597       1038   
29  0500000US51810    Virginia Beach city, Virginia     175029       2072   

   DP02_0001PE DP02_0001PM DP02_0002E DP02_0002M DP02_0002PE DP02_0002PM  ...  \
0        43754         (X)      22481       1665        51.4         3.5  ...   
1       111887         (X)      41865       2320        37.4         2.0  ...   
2        29951         (X)      17168       1482        57.3         4.4  ...   
3        31390         (X)      18059       1613        57.5         4.7  ...   
4       128920         (X)      70225       2556        54.5         2.0  ...   
5       400186         (X)     230047       4643        57.5         1.1  ...   
6        24270         (X)      14563       1297        60.0         4.9  ...   
7        32010         (X)      19368       1776        60.5         5.1  ...   
8        39521         (X)      24537       1840        62.1         4.3  ...   
9       130591         (X)      58703       3033        45.0         2.4  ...   
10       30262         (X)      17490       1546        57.8         4.2  ...   
11      133750         (X)      90154       3127        67.4         2.2  ...   
12       36244         (X)      16630       1706        45.9         4.5  ...   
13      143508         (X)      85448       3799        59.5         2.4  ...   
14       37506         (X)      19782       1646        52.7         4.5  ...   
15       30462         (X)      18095       1375        59.4         3.9  ...   
16       46427         (X)      26008       1757        56.0         3.5  ...   
17       48092         (X)      31349       1946        65.2         3.8  ...   
18       26008         (X)      14963       1293        57.5         4.7  ...   
19  

In [9]:
# searching the term "broadband" in metadata's `id` column
filtered_meta = metadata[metadata['id'].str.contains("broadband")]
filtered_meta

GEO_ID                                                 id
609   DP02_0153E  Estimate!!COMPUTERS AND INTERNET USE!!Total ho...
610   DP02_0153M  Margin of Error!!COMPUTERS AND INTERNET USE!!T...
611  DP02_0153PE  Percent!!COMPUTERS AND INTERNET USE!!Total hou...
612  DP02_0153PM  Percent Margin of Error!!COMPUTERS AND INTERNE...

The filtered dataframe's `GEO_ID` shows us which `data` labels are relevant to broadband.

Let's list the labels and inspect the relevant rows from `data`..

In [10]:
# listing the filtered `GEO_ID` column labels
broadband_vars = filtered_meta['GEO_ID'].tolist()

# viewing the list of filtered labels
broadband_vars

['DP02_0153E', 'DP02_0153M', 'DP02_0153PE', 'DP02_0153PM']

In [11]:
# subset `data` to include only columns with labels in `broadband_vars`
update_data[broadband_vars]

DP02_0153E DP02_0153M DP02_0153PE DP02_0153PM
0       38057       1723        87.0         3.0
1      103460       2365        92.5         1.4
2       25807       1681        86.2         3.8
3       24229       1757        77.2         4.3
4      115443       2812        89.5         1.7
5      382803       3539        95.7         0.6
6       21241       1293        87.5         3.0
7       26805       1677        83.7         3.1
8       33859       1666        85.7         3.2
9      112672       2952        86.3         1.8
10      27223       1677        90.0         3.5
11     127625       2436        95.4         1.2
12      33056       1661        91.2         2.6
13     137524       2159        95.8         0.8
14      32685       1696        87.1         3.0
15      24822       1670        81.5         3.7
16      42465       1419        91.5         2.1
17      45469       1466        94.5         1.8
18      23559       1098        90.6         2.7
19      65428       1931        92.1         1.6
20      79413       2063        91.4         1.8
21      47767       1933        85.9         2.5
22      26975       1370        93.2         2.2
23      60865       2938        85.8         2.6
24      73239       2546        82.9         1.9
25      31644       1221        87.7         2.3
26      73754       2573        82.1         2.4
27      32035       1907        78.5         4.3
28      29058       1309        84.0         3.3
29     162381       2694        92.8         1.1

For reference, we can create a dictionary from `metadata` to more easily reference descriptions for the columns in `data`. This will be useful for identifying columns that may interest us for analysis.

In [12]:
# creating a metadata dictionary
metadata_dict = metadata.set_index('GEO_ID').T.to_dict('list')

In [13]:
metadata_dict

{'NAME': ['Geographic Area Name'],
 'DP02_0001E': ['Estimate!!HOUSEHOLDS BY TYPE!!Total households'],
 'DP02_0001M': ['Margin of Error!!HOUSEHOLDS BY TYPE!!Total households'],
 'DP02_0001PE': ['Percent!!HOUSEHOLDS BY TYPE!!Total households'],
 'DP02_0001PM': ['Percent Margin of Error!!HOUSEHOLDS BY TYPE!!Total households'],
 'DP02_0002E': ['Estimate!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family'],
 'DP02_0002M': ['Margin of Error!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family'],
 'DP02_0002PE': ['Percent!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family'],
 'DP02_0002PM': ['Percent Margin of Error!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family'],
 'DP02_0003E': ['Estimate!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family!!With own children of the householder under 18 years'],
 'DP02_0003M': ['Margin of Error!!HOUSEHOLDS BY TYPE!!Total households!!Married-couple family!!With own children of the householder under 18 years']

This is a better visual reference.

*We could remove the first entry (del metadata['NAME']) and break the string descriptions out of their lists. Since we would not gain much-if-anything for our trouble, we will leave the dictionary as it is.*

## Helper Functions

In [14]:
# loading python's `autoreload`, to update any external module changes
%load_ext autoreload

# turning-on `autoreload`
%autoreload 2

In [15]:
# verifying the current working directory
os.getcwd()

'C:\\Users\\jamel\\myprojects\\acs-api\\notebooks'

In [16]:
# moving to the parent directory
os.chdir('..')

# verifying the current working directory
os.getcwd()

'C:\\Users\\jamel\\myprojects\\acs-api'

### Serialization

We will use the `pickle` serialization format to store objects for future use.

In [17]:
# importing the "helpers folder and contained modules as a package"
from helpers import *

Within `helpers` is the `helpers_func` package, which includes the custom `save_pickle`/`read_pickle` serialization/de-serialization modules.

In [18]:
# serializing the data dictionary
save_pickle(metadata_dict, "metadata-dict")

# serializing the list of broadband variables
save_pickle(broadband_vars, "broadband-vars")

--------------- PICKLING metadata-dict -------------------------
Saved as  metadata-dict.pkl 

--------------- PICKLING broadband-vars -------------------------
Saved as  broadband-vars.pkl 



In [19]:
# # -----uncomment cell to test-----
# # test deleting the variable
# del broadband_vars

# # attempting to view the deleted variable
# print("Result of attempting to print variable after deletion:\n")
# broadband_vars

In [20]:
# # -----uncomment cell to test-----
# print("Result of attempting to view the object after restoration:\n")

# # restoring the data dictionary from the serialized file
# broadband_vars = read_pickle("pickles/broadband-vars.pkl")
# broadband_vars

# Scrub 

# Explore

## Notebook Summary

Continue to "2-api-data" notebook.